src/lib/__tests__/SepculationEngine.test.ts

```diff
@@ -0,0 +1,254 @@
+          created: new Date(120),
+          forCommit: 'abc',
+          id: '0',
+          impact: 100,
+          triggererAaid: '123',
+          pullRequestId: 0,
+          pullRequest,
+        }),
+        requestId: '0',
+        state: 'queued',
+      }),
+      new LandRequestStatus({
+        date: new Date(120),
+        id: '1',
+        isLatest: true,
+        request: new LandRequest({
+          created: new Date(120),
+          forCommit: 'abc',
+          id: '1',
+          impact: 50,
+          triggererAaid: '123',
+          pullRequestId: 1,
+          pullRequest,
+        }),
+        requestId: '0',
+        state: 'queued',
+      }),
+    ];
+
+    mockRunning = [
+      new LandRequestStatus({
+        date: new Date(120),
+        id: '0',
+        isLatest: true,
+        request: new LandRequest({
+          created: new Date(120),
+          forCommit: 'abc',
+          id: '0',
+          triggererAaid: '123',
+          pullRequestId: 0,
+          pullRequest,
+        }),
+        requestId: '0',
+        state: 'running',
+      }),
+      new LandRequestStatus({
+        date: new Date(120),
+        id: '1',
+        isLatest: true,
+        request: new LandRequest({
+          created: new Date(120),
+          forCommit: 'abc',
+          id: '1',
+          triggererAaid: '123',
+          pullRequestId: 1,
+          pullRequest,
+        }),
+        requestId: '0',
+        state: 'running',
+      }),
+    ];
+  });
+
+  afterEach(() => {
+    jest.restoreAllMocks();
+  });
+
+  test('getAvailableSlots > should return availabled free slots', async () => {
+    jest.spyOn(StateService, 'getMaxConcurrentBuilds').mockResolvedValue(3);
+    expect(await SpeculationEngine.getAvailableSlots(mockRunning)).toBe(1);
+    expect(await SpeculationEngine.getAvailableSlots([])).toBe(3);
+  });
+
+  test('positionInQueue > should return position in queue', async () => {
+    expect(await SpeculationEngine.positionInQueue(mockQueued, mockQueued[1])).toBe(1);
+    expect(await SpeculationEngine.positionInQueue(mockQueued, mockQueued[0])).toBe(0);
+  });
+
+  test('getImpact > should return list of next queued request`s impact', async () => {
+    const impact = await SpeculationEngine.getImpact(mockQueued, 0, 2);
+    expect(impact).toEqual([100, 50]);
+  });
+
+  describe('reOrderRequest', () => {
+    test('should return false when the feature is turned off', async () => {
+      jest
+        .spyOn(StateService, 'getAdminSettings')
+        .mockResolvedValueOnce({ speculationEngineEnabled: false } as any);
+      jest.spyOn(SpeculationEngine, 'getImpact').mockResolvedValueOnce([10, 100]);
+      const reOrder = await SpeculationEngine.reOrderRequest(
+        mockRunning,
+        mockQueued,
+        mockQueued[0],
+      );
+
+      expect(SpeculationEngine.getImpact).not.toHaveBeenCalled();
+      expect(reOrder).toBe(false);
+    });
+    describe('when feature is turned on', () => {
```

Comment:
I think it'd be better to use these `describe` blocks to distinguish between successful and unsuccessful reorder cases. It reduces cognitive overhead and means we can more easily identify missing tests such as there not being a test for a successful re-order with 3 free slots.

The feature being turned off would just be a 'unsuccessful reorder case', it doesn't warrant being its own describe imo.